src/librustc_typeck/check/cast.rs

```diff
@@ -281,35 +281,12 @@ impl<'a, 'gcx, 'tcx> CastCheck<'tcx> {
                                 .emit();
             }
             CastError::SizedUnsizedCast => {
-                let mut err = type_error_struct!(
-                    fcx.tcx.sess,
-                    self.span,
-                    self.expr_ty,
-                    E0607,
-                    "cannot cast thin pointer `{}` to fat pointer `{}`",
-                    self.expr_ty,
-                    fcx.ty_to_string(self.cast_ty)
-                );
-                if fcx.tcx.sess.opts.debugging_opts.explain
-                    && !fcx.tcx.sess.parse_sess.span_diagnostic
-                        .code_emitted(&err.get_code().unwrap()) {
-                    err.help(
-                        "Thin pointers are \"simple\" pointers: they are purely a reference to a
-memory address.
-
-Fat pointers are pointers referencing \"Dynamically Sized Types\" (also
-called DST). DST don't have a statically known size, therefore they can
-only exist behind some kind of pointers that contain additional
-information. Slices and trait objects are DSTs. In the case of slices,
-the additional information the fat pointer holds is their size.
-
-To fix this error, don't try to cast directly between thin and fat
-pointers.
-
-For more information about casts, take a look at The Book:
-https://doc.rust-lang.org/book/first-edition/casting-between-types.html");
-                }
-                err.emit();
+                use structured_errors::{SizedUnsizedCastError, StructuredDiagnostic};
+                SizedUnsizedCastError::new(&fcx.tcx.sess,
```

Comment:
Personally, I'd rather something with labels I think, like:

```
SizedUnsizedCastError {
    span: self.span,
    thin_ty: self.expr_ty,
    fat_ty: fcx.ty_to_string(self.cast_ty) // although if we could do `self.cast_ty` it'd be even better
}.emit(fcx.tcx);
```

These labels would then also appear in the templating logic.

Reply:
I'm thinking that the templating could look something like the following:

```
diagnostic_template!(
Any independent text will be notes.

Such as this one and the previous one.

Here is where a span would appear:
$secondary_span the label for the span immediately after

I haven't thought about how to deal with suggestions.
)
```
```
error[E0XXX]: error message outside of template
  --> $FILE:21:9
   |
21 |         foo();
   |         ^^^^^
   = note: Any independent text will be notes.

          Such as this one and the previous one.
note: Here is where a span would appear:
  --> $FILE:21:9
   |
21 |         foo();
   |         ^^^^^ the label for the span immediately after
   = note: I haven't thought about how to deal with suggestions.
```

In the background it would convert the template string into calls to `err.*()`.

How does that sound?

Reply:
Interesting. Quite different than what I had in mind, but I like it as a start, and maybe more practical than some of my more grandiose ideas =)

I think one key question is whether the `--teach` output would still have the "basic structure" of the messages that we've been giving (as in this PR), or whether it would change to something more essay like. 

I think I had initially imagined that, in `--teach` mode, we would change "our look" somewhat drastically. e.g (using E0010 as an example):

```
error[E0010]: boxed values cannot be used in statics
  --> $FILE:21:9
  |
21 |         const CON : Box<i32> = box 0;
   |                                ^^^

The value of statics and constants must be known at compile time, and they live
for the entire lifetime of a program. Creating a boxed value allocates memory on
the heap at runtime, and therefore cannot be done at compile time.
```

For more complex examples, there might be multiple snippets:

```
error[E0010]: shadowed constant `CON`
  --> $FILE:21:9
  |
21 |         const CON : Box<i32> = box 0;
   |               ^^^

This code is illegal. The problem is that the const CON is shadowed by another constant declared earlier, also named `CON`:

  --> $FILE:18:9
  |
18 |         const CON : Box<i32> = box 22;
   |               ^^^
```

Reply:
The only reason I wasn't considering moving in that direction is because I though we wouldn't want to present the same explanation over and over again, so mixing the current diagnostics with the teach diagnostics would be a bit jarring. But if we don't mind either mixing both styles or repeating walls of text, then [I don't see why we wouldn't do it](https://github.com/rust-lang/rfcs/blob/master/text/1644-default-and-expanded-rustc-errors.md#expanded-error-messages).

Reply:
Also, I think that the differences in output could act like a stylesheet, where the `DiagnosticBuilder` is still created in the same way, but when emitting it to the output we change how things are presented depending on wether we're in `teach` mode or not, by removing the left margin on `note`s, line wrapping, how we deal with suggestions, maybe [even highlighting code](https://github.com/rust-lang/rust/pull/45752) based on this.